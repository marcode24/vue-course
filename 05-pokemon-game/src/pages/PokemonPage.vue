<template>
  <h1 v-if="!pokemon"> Espere por favor... </h1>
  <div v-else>
    <h1>¿Quién es este pokémon?</h1>
    <PokemonPicture :pokemonId="pokemon.id" :showPokemon="showPokemon"/>
    <PokemonOptions :pokemons="pokemonOptions" @selection="checkAnswer"/>
    <template v-if="showAnswer">
      <h2 class="fade-in">{{ message }}</h2>
      <button @click="newGame">Nuevo juego</button>
    </template>
  </div>
</template>

<script>

import PokemonOptions from '@/components/PokemonOptions';
import PokemonPicture from '@/components/PokemonPicture';
import getPokemonOptions from '@/helpers/getPokemonOptions';

export default {
  name: 'PokemonPage',
  components: {
    PokemonOptions,
    PokemonPicture
  },
  data() {
    return {
      pokemonOptions: [],
      pokemon: null,
      showPokemon: false,
      showAnswer: false,
      message: ''
    }
  },
  methods: {
    async mixPokemonsArray() {
      this.pokemonOptions = await getPokemonOptions();
      const randomIndex = Math.floor(Math.random() * this.pokemonOptions.length);
      this.pokemon = this.pokemonOptions[randomIndex];
    },
    checkAnswer(pokemonId) {
      if(pokemonId === this.pokemon.id) {
        this.message = '¡Correcto!';
      } else {
        this.message = `¡Incorrecto! El pokémon correcto es ${this.pokemon.name}`;
      }
      this.showPokemon = true;
      this.showAnswer = true;
    },
    newGame() {
      this.showPokemon = false;
      this.showAnswer = false;
      this.pokemon  = null;
      this.pokemonOptions = [];
      this.mixPokemonsArray();
    }
  },
  mounted() {
    this.mixPokemonsArray();
  }
}
</script>

<style>

</style>