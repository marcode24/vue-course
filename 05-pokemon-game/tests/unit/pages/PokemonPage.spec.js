import { shallowMount, mount } from '@vue/test-utils'
import PokemonPage from '@/pages/PokemonPage'
import { pokemons } from '../mocks/pokemons.mock';

describe('PokemonPage Component', () => {
  let wrapper;
  let mixPokemonsArraySpy;

  beforeEach(() => {
    mixPokemonsArraySpy = jest.spyOn(PokemonPage.methods, 'mixPokemonsArray');
    wrapper = shallowMount(PokemonPage);
  });

  test('Debe hacer match con el snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  test('Debe de llamar al mixin de "getPokemons" al montar el componente', () => {
    expect(mixPokemonsArraySpy).toHaveBeenCalled();
  });

  test('Debe de hacer match con el snapshot cuando se cargan los pokemons', () => {
    const wrapper = mount(PokemonPage, {
      data() {
        return {
          pokemonOptions: pokemons,
          pokemon: pokemons[0],
          showPokemon: false,
          showAnswer: false,
          message: ''
        };
      }
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  test('Debe de mostrar los componentes de PokemonPicture y PokemonOptions', () => {
    const wrapper = shallowMount(PokemonPage, {
      data() {
        return {
          pokemonOptions: pokemons,
          pokemon: pokemons[0],
          showPokemon: false,
          showAnswer: false,
          message: ''
        };
      }
    });
    const picture = wrapper.find('pokemon-picture-stub');
    const options = wrapper.find('pokemon-options-stub');

    expect(picture.exists()).toBeTruthy();
    expect(options.exists()).toBeTruthy();

    expect(picture.attributes('pokemonid')).toBe('1');
    expect(options.attributes('pokemons')).toBeTruthy();
  });

  test('Prueba con checkAnswer', async () => {
    const wrapper = shallowMount(PokemonPage, {
      data() {
        return {
          pokemonOptions: pokemons,
          pokemon: pokemons[0],
          showPokemon: false,
          showAnswer: false,
          message: ''
        };
      }
    });

    await wrapper.vm.checkAnswer(5);
    expect(wrapper.find('h2').exists()).toBeTruthy();
    expect(wrapper.vm.showPokemon).toBeTruthy();
  });
})