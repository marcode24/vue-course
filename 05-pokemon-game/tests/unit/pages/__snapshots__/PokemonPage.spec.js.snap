// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonPage Component Debe de hacer match con el snapshot cuando se cargan los pokemons 1`] = `
<div>
  <h1>¿Quién es este pokémon?</h1>
  <div class="pokemon-container"><img class="hidden-pokemon" src="https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/dream-world/1.svg" alt="pokemon-image"></div>
  <div class="options-container">
    <ul>
      <li>bulbasaur</li>
      <li>ivysaur</li>
      <li>venusaur</li>
      <li>charmander</li>
    </ul>
  </div>
  <!--v-if-->
</div>
`;

exports[`PokemonPage Component Debe hacer match con el snapshot 1`] = `<h1> Espere por favor... </h1>`;
